### ISSUE TITLE
No way to specify ACL for IPv6 subnets

### ISSUE THREAD

**Issue by jimevans** [civetweb/civetweb#591]
When compiling with USE_IPV6, no values can be specified for the access control list for IPv6 subnets. In my implementation, I want to listen on a port on both IPv4 and IPv6 addresses, but only allow connection from the local machine. Therefore, the access control list I want is something like:

    -0.0.0.0/0,-::/0,+127.0.0.1,+::1

Attempting to use such an ACL as follows does not work:

    mg_callbacks callbacks = {};
    const char* options[] = {
        "listening_ports", "+80",
        "access_control_list", "-0.0.0.0/0,-::/0,+127.0.0.1,+::1",
        NULL };
    mg_context* context = mg_start(&callbacks, NULL, options);

**Comment by bel2125** [comment]
Currently there is no way to use IPv6 in an acl - the acl implementation still uses `uint32_t` to store IP addresses. I know this issue, but planned to address it only after changing to a more sophisticated pattern matching (#499).

But, for what you want, you could just bind only to the loopback adapter instead of using an acl.
```C
const char* options[] = {
    "listening_ports", "127.0.0.1:80,[::1]:80",
    /* no access_control_list */
    NULL };
```

**Comment by jimevans** [comment]
I suppose you're right, at that. Feel free to do with this issue what you will. If you'd like to leave it open as a marker for future work, that's fine. If you want to close it as resolved because there's a workaround, and my particular use case is solved, I'm perfectly happy with that too.

**Comment by bel2125** [comment]
Postponing until a more powerful pattern matching is available (see https://github.com/civetweb/civetweb/issues/499).
Furthermore, it needs to be evaluated if IP-pattern based access control is reasonable at all for IPv6.

**Comment by nfrmtkr** [comment]
Today I was about to log this same issue, but saw its already addressed. 
The workaround for loopback device is fine if you have a IPv4 and IPv6 mixed network stack. With a IPv6-only network stack you cant use it. Well, most probably you will hit IPv6-only configurations in rare cases, but it will change in future.
I wouldn't make this issue depend on #499. for IPv6 support maybe a lits of IPv6 address would be enough. One use case is in configurations where another http server (e.g. Apache) is used as proxy to ciwetweb http server. With the ACL you will identify a particular host that is allowed to access the civetweb http. In this case a IPv6 address or list is sufficient.

**Comment by bel2125** [comment]
> The workaround for loopback device is fine if you have a IPv4 and IPv6 mixed network stack. With a IPv6-only network stack you cant use it.


Why not? As long as you want to limit access to localhost, you should still be able to use `listening_ports [::1]:80".
Or do you want to limit access to a specific IPv6 address not belonging to the local machine?

**Comment by nfrmtkr** [comment]
The use case listening on a IPv6 network device, but allowing access of one remote IPv6 machine, only.
For instance a machine running a public Apache HTTP server acting as proxy to a machine on private network running civetweb.

**Comment by bel2125** [comment]
Up to now all private networks I encountered had private IPv4 addresses (10.x.y.z) - and I got no indication this is going to change any time soon. I also heard various reasons for this - easier configuration, more experience of the maintanance staff, more stable, ... a network security consultant told me (2 years ago) security configuration is far more complicated with IPv6 than with IPv4 and some components are not IPv6 ready from the security point of view. Of course, this may change in the future.
Do you see any signs this is starting to change ***now***?

If this became a more relevant use case, we can certainly rank it higher.

**Comment by nfrmtkr** [comment]
Well, today it's mor for completeness. If we support listening on IPv6 network device then we should also support ACL on it. However, similar things admins may achive by configuring the firewall, isn't it?

**Comment by bel2125** [comment]
Fixed by #886
